R/cube-dims.R

```diff
@@ -144,29 +144,84 @@ setMethod("[", "CubeDims", function (x, i, ...) {
+#' @keywords internal
+getDimTypes <-  function (x) {
+    if (inherits(x, "CrunchCube")) {
+        x <- x@dims
+    }
+
+    what_dim_is_it <- function (one_var, array_aliases) {
+        dim_type <- type(one_var)
+        
+        if (alias(one_var) %in% array_aliases) {
+            # we are in an array, we need to figure out if this is a multiple
+            # response or not
+            array_cat_dim <- vars[aliases(vars) == alias(one_var) & types(vars) == "categorical"]
+
+            # if we can't find the matching categories dimension we might have a
+            # subset cube, so simply return the dim_type un-identified (this
+            # might could actually just be "ca_items")
+            if (length(array_cat_dim) < 1) {
+                return(dim_type)
+            }
+            
+            # if this is a variable crossed by itself, then array_cat_dim will
+            # actually have two copies of the categories dimension. We take the
+            # first one becasue it should be identical to all the others. If it
+            # isn't this might produce weird results. 
+            # TODO: Investigate checking by ID
+            array_cats <- categories(array_cat_dim[[1]])
+            
+            # if we meet these conditions, we are actually a multiple response
+            # and should label ourself as such.
+            # Unlike the strict is.3vl, this doesn't compare cat names because
+            # they've already been munged to TRUE/FALSE
+            is.MR <- length(array_cats) == 3 &&
+                setequal(ids(array_cats), c(-1, 0, 1)) &&
+                sum(is.selected(array_cats)) == 1 &&
+                sum(is.na(array_cats)) == 1
+            
+            if (is.MR && dim_type == "subvariable_items") {
+                dim_type <- "mr_items"
+            }
+            if (is.MR && dim_type == "categorical") {
+                dim_type <- "mr_selections"
+            }
+            if (!is.MR && dim_type == "subvariable_items") {
+                dim_type <- "ca_items"
+            }
+            if (!is.MR && dim_type == "categorical") {
+                dim_type <- "ca_categories"
+            }
         }
-        return(maybe)
+        
+        return(dim_type)
     }
-    vars <- variables(dims)
-    # We only need to check if the categories are the magical Selected
-    # categories if there is an MR somewhere with the same alias
-    MRaliases <- aliases(vars)[types(vars) == "subvariable_items"]
-
-    # determine which dimensions are selected MR dimensions
-    selecteds <- mapply(is.it, x=index(vars), dim=dims@.Data,
-                        MoreArgs=list(MRaliases=MRaliases))
-    names(selecteds) <- dims@names
-    return(selecteds)
+  
+    vars <- variables(x)
+    array_aliases <- aliases(vars)[types(vars) == "subvariable_items"]  
+    # vapply becuase lapply on a VariableCatalog yields lists (not tuples)
+    out <- unlist(vapply(vars, what_dim_is_it, character(1), array_aliases))
```

Comment:
Why `unlist()` if you're returning a character vector from vapply (which also seems to call into question the comment on L222: it sounds like vapply is actually what you want anyway).

Reply:
Good catch, it was an unnecessary holdover from using `lapply` there before.